org/mozilla/jss/nss/SSLFDProxy.java

```diff
@@ -46,4 +49,12 @@ protected synchronized void releaseNativeResources() throws Exception {
             }
         }
     }
+
+    public int invokeCertAuthHandler() {
+        if (handler == null) {
+            return -5994;
```

Comment:
I'm thinking that if `invokeCertHandler` raises an exception, we should catch that in the JNI code:

https://github.com/dogtagpki/jss/pull/579/files#diff-a977b6f7e7ac81cab2721a9b5333e64fR357
```c
    ret = (*env)->CallIntMethod(env, sslfd_proxy, certAuthHandlerMethod);
    PR_SetError(ret, 0);
```

Prior to `PR_SetError`, check if an exception occurred and set it there instead. Thoughts?

Reply:
That'll work, but are we going to throw a generic NPE or something more specific? Another option is to get the handler from the proxy, then call `handler.check()` directly. Something like this but using JNI:
```
handler = proxy.getHandler();
if (handler == null) {
    PR_SetError(-5994, 0);
    return;
}

handler.check(proxy);
```

Reply:
No, the NPE would be thrown by `invokeCertAuthHandler` when `handler.check(this)` fails because `handler == null`. We'd check whether _any_ exception occurs and make sure to call `PR_SetError(PR_UNKNOWN_ERROR, 0)` when `(*env)->ExceptionOccurred(env) != NULL`. We wouldn't clear the exception, so when we get back to Java (on the other side of this NSS callback), it'd still have the Java exception _and_ it'd indicate to NSS that cert isn't trusted.

We could do the same and allow `check` to throw an exception too I suppose. Though, this wouldn't work well with `run()` (since it cannot throw an exception either way).

Thoughts?

Reply:
Not sure I understand. You said "when we get back to Java ... it'd still have the Java exception", could you show me how Java would catch the exception?

Regardless, the original code was fine. I was just not clear what `-5994` means, so if we can create a constant, or just add a comment describing the value that should be fine.

Reply:
So the call stack looks like this:

(during initialization, setup the sync cert handler)

 1. `SSLEngine.wrap()`
 2. `SSLEngine.updateHandshakeState()`
 3. `SSL.ForceHandshake(ssl_fd)` =~ `SSL_ForceHandshake(PRFileDesc *fd)` (Java -> NSS via JNI)
 4. ... nss internals ...
 5. NSS invokes `ss->authCertificate()` -- we've replace this with our synchronous cert auth handler during setup. So this is equivalent to invoking `JSSL_SSLFDSyncCertAuthCallback`. 
 6. `SSLFDProxy.invokeCertAuthHandler()` 
 7. `SSLFDProxy.handler.check(this)`
 8. ... user code in Java &c ...

Now, say user code raises an exception or `handler == null`. In `JSSL_SSLFDSyncCertAuthCallback`, we can use `(*env)->ExceptionOccurred(env)` to ensure that `ret = PR_UNKNOWN_ERROR`. This will ensure that NSS knows an exception occurred and that the cert shouldn't be trusted, i.e., handshake should terminate. It'll eventually unravel until `SSL.ForceHandshake(ssl_fd)` returns, at which point the JNI layer will exit and we're back in pure Java land.

When exiting from JNI layer, the exception raised from somewhere inside `SSLFDProxy.invokeCertAuthHandler` will still be thrown. We'll be able to either handle it (most likely, wrap it in a `SSLException` and re-trhow) or let it continue unrolling back to user code (if it is a `RuntimeException`). Likely this'd go back all the way to Tomcat/... and the app would quit handshaking, though something should've logged the exception.

Reply:
JFTR, I took this approach. It means the `NullPointerException` will be preserved (and we won't artificially avoid it), and we can use the C `PR_UNKNOWN_ERROR` constant in the `PR_SetError` call.